src/layers.c

```diff
@@ -118,45 +118,70 @@ handle_output_destroy(struct wl_listener *listener, void *data)
+ */
+static bool
+has_precedence(struct seat *seat, enum zwlr_layer_shell_v1_layer layer)
+{
+	if (!seat->focused_layer) {
+		return true;
+	}
+	if (!focused_layer_has_exclusive_interactivity(seat)) {
+		return true;
+	}
+	if (layer >= seat->focused_layer->current.layer) {
+		return true;
+	}
+	return false;
+}
+
+void
+layer_try_set_focus(struct lab_layer_surface *layer)
 {
 	struct wlr_layer_surface_v1 *layer_surface = layer->scene_layer_surface->layer_surface;
 	struct seat *seat = &layer->server->seat;
 
-	if (layer_surface->current.keyboard_interactive
-			&& layer_surface->current.layer >= ZWLR_LAYER_SHELL_V1_LAYER_TOP) {
-		/*
-		 * Give keyboard focus to surface if
-		 * - keyboard-interactivity is 'exclusive' or 'on-demand'; and
-		 * - surface is in top/overlay layers; and
-		 * - currently focused layer has a lower precedence
-		 *
-		 * In other words, when dealing with two surfaces with
-		 * exclusive/on-demand keyboard-interactivity (firstly the
-		 * currently focused 'focused_layer' and secondly the
-		 * 'layer_surface' for which we're just responding to a
-		 * map/commit event), the following logic applies:
-		 *
-		 * | focused_layer | layer_surface | who gets keyboard focus |
-		 * |---------------|---------------|-------------------------|
-		 * | overlay       | top           | focused_layer           |
-		 * | overlay       | overlay       | layer_surface           |
-		 * | top           | top           | layer_surface           |
-		 * | top           | overlay       | layer_surface           |
-		 */
-
-		if (!seat->focused_layer || seat->focused_layer->current.layer
-				<= layer_surface->current.layer) {
+	switch (layer_surface->current.keyboard_interactive) {
+	case ZWLR_LAYER_SURFACE_V1_KEYBOARD_INTERACTIVITY_EXCLUSIVE:
+		wlr_log(WLR_DEBUG, "interactive-exclusive '%p'", layer_surface);
+		if (has_precedence(seat, layer_surface->current.layer)) {
 			seat_set_focus_layer(seat, layer_surface);
 		}
-	} else if (seat->focused_layer
-			&& !seat->focused_layer->current.keyboard_interactive) {
-		/*
-		 * Clear focus if keyboard-interactivity has been set to
-		 * ZWLR_LAYER_SURFACE_V1_KEYBOARD_INTERACTIVITY_NONE
-		 */
-		seat_set_focus_layer(seat, NULL);
+		break;
+	case ZWLR_LAYER_SURFACE_V1_KEYBOARD_INTERACTIVITY_ON_DEMAND:
+		wlr_log(WLR_DEBUG, "interactive-on-demand '%p'", layer_surface);
+		if (!focused_layer_has_exclusive_interactivity(seat)) {
+			seat_set_focus_layer(seat, layer_surface);
+		}
+		break;
+	case ZWLR_LAYER_SURFACE_V1_KEYBOARD_INTERACTIVITY_NONE:
+		wlr_log(WLR_DEBUG, "interactive-none '%p'", layer_surface);
+		if (seat->focused_layer == layer_surface) {
+			wlr_log(WLR_DEBUG, "unset focus '%p'", layer_surface);
+			/*
+			 * TODO: consider transferring focus other layer-shell
+			 * clients with exclusive focus (that in the case of
+			 * multiple clients we could have stole it from and
+			 * arguably should give it back to).
+			 */
+			seat_set_focus_layer(seat, NULL);
```

Comment:
Should we do this in this PR? `seat_set_focus_layer(seat, NULL)` gives keyboard focus to the topmost usual window currently. I assume we should do something like this instead:
- exclusive-overlay-and-top
- on_demand-overlay-and-top ?
- topmost_view ?
- exclusive-bottom-and-background
- on_demand-bottom-and-background

and break on the first match. Not quite sure about `on_demand-overlay-and-top` before `topmost_view` or the other way around.

Reply:
I've added `try_to_focus_next_layer()`

It merely does the following (in order of precedence):

- Give focus to last added overlay/top layer-shell client with exclusive interactivity. I think this is quite an edge case in practice, but seems tidy. For example if you run two bemenu-run instances (I'm not sure why you would, but still!) when the first one exits, the second one gets keyboard focus.

- Give focus to topmost toplevel (=view) if one exists. We actually already do this, so am just noting it here for completeness. https://github.com/labwc/labwc/blob/a5fcbfaf720fc6c2486d850012215102e90e04bf/src/seat.c#L664

> maybe some previously-focused?

Well, if we're talking `exclusive` clients, presumable they had focus before it was stolen from them, so logic above just give it back to them.

They've been inserted at the back of the list in wlroots/types/scene/wlr_scene.c (`wl_list_insert(parent->children.prev, &node->link);` so iterating in reverse gives focus to last added client first.

https://gitlab.freedesktop.org/wlroots/wlroots/-/blob/master/types/scene/wlr_scene.c?ref_type=heads#L72

> on_demand-overlay-and-top?

I haven't implemented this. I think it would only make sense if such a client had focus and then it was stolen from them by another layer-shell client. Otherwise I think it's more intuitive to give it to the topmost toplevel. I suspect it would be hard to implement too and would probably require a list to keep track of clients from whom keyboard interactivity was stolen. Suggest we leave that for a separate PR - if ever.

> exclusive-bottom-and-background? 

Decided not to implement. It would be dead easy to do, but not sure what the use-case is. I don't mind either way though.

> on_demand-bottom-and-background

Same argument as on-demand-overlay-and-top.

> Not quite sure about on_demand-overlay-and-top before topmost_view or the other way around.

If you take the case of a panel (which could be in `top` or `bottom`) the question is really whether it would be better to give focus to the topmost toplevel or the panel.

What we could consider (but I prefer do consider this in a separate PR) is to close popups when focus is stolen. That would feel quite 'normal' and reduce the sense that focus should be given back at some point.

> The logic could theoretically be moved to seat_set_focus_layer(seat, NULL)

Looks like `seat_input_inhibit()` wants to clear focus without fallbacks (i.e. need to keep `seat_set_focus_layer()` as is in support of that)

> and should likely also be called on unmap if the layer was focused at that point.

Done.